tests/models/encoder_decoders/wikitables_semantic_parser_test.py

```diff
@@ -25,6 +25,105 @@ def setUp(self):
+                                                                     [-1, -1]]]))
+
+    def test_get_type_vector(self):
+        worlds, num_entities = self.get_fake_worlds()
+        tensor = Variable(torch.LongTensor([]))
+        type_vector, _ = self.model._get_type_vector(worlds, num_entities, tensor)
+        # Verify that both types are present and padding used for non existent entities.
+        assert torch.equal(type_vector.data, torch.LongTensor([[[1, 0],
+                                                                [1, 0],
+                                                                [0, 1],
+                                                                [0, 1],
+                                                                [0, 0],
+                                                                [0, 0]],
+                                                               [[1, 0],
+                                                                [0, 1],
+                                                                [0, 0],
+                                                                [0, 0],
+                                                                [0, 0],
+                                                                [0, 0]]]))
+
+    def test_get_linking_probabilities_valid_probability_distribution(self):
+        worlds, num_entities = self.get_fake_worlds()
+        tensor = Variable(torch.FloatTensor([]))
+        _, entity_type_dict = self.model._get_type_vector(worlds, num_entities, tensor)
+        questions = Variable(torch.LongTensor([[14, 40],
+                                               [78, 0]]))
+        # (batch_size, num_entities, num_question_tokens)
+        linking_scores = [[[0.6218093, 0.74813986],
+                           [0.82922572, 0.85793620],
+                           [0.11054164, 0.83972567],
+                           [0.72952926, 0.52519375],
+                           [0.34569067, 0.73687404],
+                           [0.30000000, 0.55145717]],
+                          [[0.61507487, 0.89550155],
+                           [0.74650306, 0.95761484],
+                           [0.51626843, 0.22729224],
+                           [0.52449512, 0.06206334],
+                           [0.24870688, 0.92492068],
+                           [0.8245101, 0.20058638]]]
+        linking_scores = Variable(torch.FloatTensor(linking_scores))
+        question_mask = get_text_field_mask({'questions': questions}).float()
+
+        # (batch_size, num_question_tokens, num_entities + 1)
+        entity_probability = self.model._get_linking_probabilities(worlds, linking_scores, question_mask,
+                                                                   entity_type_dict, tensor)
+
+        # The following properties in entity_probability are tested for by true_probability:
+        # It has probability 0.0 at the 0th entity index for the null entity.
```

Comment:
This isn't actually correct - we don't want a 0 _probability_ for the null entity, we want a zero _logit_ for the null entity.  If you look at the right column on page four of the paper, you can see that `s(null, i) = 0`, but `exp(0) = 1` - this allows the model to select the null entity if it doesn't think there's a good match.  Having negative values for some of the scores above would help show that this is working.

Also, I think the value for the null entity here doesn't matter, right?  Because that entity shows up in several normalizations, it's not clear what you should put in the table.  But the value in the table doesn't ever get used, so it doesn't really matter.  You could even just remove it, and return a tensor of shape `(batch_size, num_question_tokens, num_entities)`.

Reply:
So should I just append a 0 to cell_type_scores and row_type_scores so it will be used in the softmax? Then return a tensor of shape (batch_size, num_question_tokens, num_entities)?

In the weighted sum equation the null entity embedding doesn't seem to be included, does that mean I should remove it from forward() ?

Reply:
Yeah, we want a non-masked 0 added to each of the softmaxes for the null entity.  And then you can either return a tensor of shape `(batch_size, num_question_tokens, num_entities + 1)`, and include a row for the null entity embedding in your weighted sum, or return a tensor of shape `(batch_size, num_question_tokens, num_entities)`, and don't have a row for the null entity.  I don't think it makes a big difference, though the first option is probably simpler.